### PR TITLE
new THREE.ShadowMesh

### PR DESCRIPTION
**Pull request by erichlof**
This PR adds a simple ShadowMesh class.  Here's a demo of the new class: <pr>
[ShadowMesh Demo](http://erichlof.github.io/ShadowMesh-Demo/ShadowMesh-Demo.html) 

Step-by-step usage patterns (and for future documentation) can be found here:
https://github.com/erichlof/ShadowMesh-Demo

@mrdoob 
The API is much cleaner now - thanks for the suggestion!  Hopefully everything is in order.

Thanks! :)
-Erich

### CODE REVIEW

File: src/objects/ShadowMesh.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
+// File:src/objects/ShadowMesh.js
+
+/**
+ * @author erichlof /  http://github.com/erichlof
+ */
+
+// a shadow Mesh that follows its parent Mesh in the scene, but is confined to a single plane.
+
+THREE.ShadowMesh = function ( parentMesh ) {
+
```
Comment (WestLangley) [inline src/objects/ShadowMesh.js]: I think the argument should be `mesh` or `object` -- as it is with the helpers.

File: src/objects/ShadowMesh.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		sme[ 3 ]  = - lightPosition4D.w * plane.normal.x;
+		sme[ 7 ]  = - lightPosition4D.w * plane.normal.y;
+		sme[ 11 ] = - lightPosition4D.w * plane.normal.z;
+		sme[ 15 ] = dot - lightPosition4D.w * -plane.constant;
+	
+		this.matrix.identity();
+		this.matrix.multiply( shadowMatrix );
+		this.matrix.multiply( this.parentMeshMatrix );
+
```
Comment (WestLangley) [inline src/objects/ShadowMesh.js]: Instead of the previous 3 lines:

		this.matrix.multiplyMatrices( shadowMatrix, this.parentMeshMatrix );

**Comment by WestLangley** [comment]
1. See if you can get it to work if the mesh is not a child of the scene, but is a child of another object.

2. This may be useful for `CanvasRender`, but the computed transform is non-invertible. This is a problem.

**Comment by erichlof** [comment]
@WestLangley 
I made the changes you requested. Hopefully everything looks good now. As to following an object that is not a child of the scene,  I don't see any reason it couldn't.  I'll have to look into it because it involves some matrix manipulations that I'm unfamiliar with - but I will try to get a small demo of a group being shadowed or something.  
@mrdoob And westLangley
In the meantime,  I discovered a cool way to make the shadows appear to move smoothly from one plane to another as in a small room with corners.  Here is a demo showing off this capability - <br> http://erichlof.github.io/ShadowMesh-Demo/ShadowMesh-MultipleInstancesDemo.html

Enjoy,  I'll let you know when I figure the group stuff out :-)

**Comment by mrdoob** [comment]
>  Here is a demo showing off this capability: <br> [ShadowMesh-MultipleInstancesDemo](http://erichlof.github.io/ShadowMesh-Demo/ShadowMesh-MultipleInstancesDemo.html) 

That looks pretty cool!

**Comment by mrdoob** [comment]
Do you mind adding a example too? O:)

**Comment by erichlof** [comment]
@mrdoob 
Thanks!  Yes I'll be glad to add an example - maybe a similar one to the one above,  but with switchable directional/point lights like in the first example? 
BTW I have successfully implemented shadows for objects that are not children of the scene,  per WestLangley's request.  I finally got the right combo of matrix multiplications necessary to handle all situations.  :D 
Now the user will be able to pass into the ShadowMesh's constructor an optional reference to the group parent or higher ancestor and all children will be shadowed appropriately.  I'm away from my desktop right now but I will post this update to this PR as well as a proof-of-concept group/children demo. Then when you and West approve everything I will submit a final generic ShadowMesh demo that could be added to the examples folder if you wish.  :)

**Comment by WestLangley** [comment]
> Now the user will be able to pass into the ShadowMesh's constructor an optional reference to the group parent or higher ancestor and all children will be shadowed appropriately.

I think you are making it more complicated than it needs to be. Just do this:

    this.meshMatrix = mesh.matrixWorld;

Please do not forget to investigate my point 2. above.

**Comment by erichlof** [comment]
Hi @mrdoob and @WestLangley ,
Here is a proof-of-concept demo showing shadowing of objects that are not children of the Scene.  Instead they are children of a THREE.Group parent: 
[ShadowMeshGroup-Demo](http://erichlof.github.io/ShadowMesh-Demo/ShadowMeshGroup-Demo.html)  <br>
@WestLangley ,
You were right, I over-complicated things initially (although it was a nice lesson in matrix manipulations :D ).  Now everything just works, whether the objects are children of the Scene or not.
Thank you for your elegant solution! - I had forgotten about the matrixWorld. 
As to point 2 above, I'm sorry but I'm not sure what you mean by non-invertable.  I am not an expert in matrices.  But if you clarify what the error would be, I will research it.

**Comment by WestLangley** [comment]
Try `CanvasRenderer` and `SoftwareRenderer`. See if you can track down console errors that appear.

**Comment by erichlof** [comment]
Ok will do! :-)

**Comment by erichlof** [comment]
@WestLangley 
Hi I tried with both SoftwareRenderer and CanvasRenderer but could not produce any errors with the simple demo code above.  Here's a live link to both:
[ShadowMesh-SoftwareDemo](http://erichlof.github.io/ShadowMesh-Demo/ShadowMesh-SoftwareDemo.html)  <br>
[ShadowMesh-CanvasDemo](http://erichlof.github.io/ShadowMesh-Demo/ShadowMesh-CanvasDemo.html)  <br>

The only time matrix inverse errors appear is when I go in and set planeConstant to 0.0 or lightPosition4D.w to 0.0.  This happens across all renderers, not just these 2.  

If you want to totally avoid these errors, I could add a simple check in the ShadowMesh constructor to make sure that both of these values are greater than 0.  If they are == 0, an arbitrarily small value could be added to them, like 0.000001, which in my tests avoid the errors altogether.  The errors themselves happen in the first place because the matrix.getInverse function is trying to divide by these values at some point, which results in infinity if either are == 0.

More importantly in my opinion, these demos don't exactly work in either Software or Canvas with the light .w value set to a pointLight value.  I tried for hours without success.  (The directional light works however for both of these renderers).  There's also some other issues such as the ground plane clipping in canvas renderer and z fighting between shadow and ground.  In the Software renderer, the shadows appear black - no transparency.  I'm not familiar enough with either of these renderers to go through their source and fix the projection/matrix problems.  Plus I have always targeted three.js 's WebGL renderer (which is the way forward in general right?).

Is it important that this shadow mesh works across all renderers (backwards compatible) , or is this PR good enough to merge now, to be used with WebGL renderer, where it works as expected?

**Comment by WestLangley** [comment]
> Hi I tried with both SoftwareRenderer and CanvasRenderer but could not produce any errors with the simple demo code above.

Hmm. I replaced `WebGLRenderer` with `CanvasRenderer` in your original demo and got errors because your transform had zero determinant. I am surprised you are not able to reproduce that. I was wondering why one renderer was generating these errors and the other was not.

> Is it important that this shadow mesh works across all renderers

No. I was just hoping your research would help explain what was going on.

**Comment by erichlof** [comment]
@WestLangley 
The only way I was able to get errors was setting certain values to 0 - and this was consistent across all renderers.  I substituted CanvasRenderer and SoftwareRenderer for WebGLRenderer with one line of code as you did, but I didn't get any errors out of the box.  
Would you like me to add the check for planeConstant == 0 and LightPosition4D.w == 0 in the ShadowMesh constructor to keep unsuspecting users from seeing possible errors?

It would be pretty neat if CanvasRenderer and SoftwareRenderer supported this feature - I tried, I really did.  But I don't know enough about their inner workings to pinpoint/fix the issue with a directional light (maximum-skewed) shadow.

**Comment by WestLangley** [comment]
A suggestion I would make would be not to use the word "parent" as your are using it. The shadow is not a child of the mesh. Instead do this:

    this.meshMatrix = mesh.matrixWorld;

For the record, you are setting the object matrix to be a projection matrix instead of an affine one, and the determinant of your object matrix is close to zero -- if not zero. Part of the rendering logic is to create a normal matrix from the object matrix, and that step requires an inversion. I do not understand all the ramifications of doing what you are doing within the framework of three.js, but my guess is it is going to lead to problems eventually.

**Comment by mrdoob** [comment]
> Is it important that this shadow mesh works across all renderers (backwards compatible) , or is this PR good enough to merge now, to be used with WebGL renderer, where it works as expected?

The PR looks good. Just missing a simple example :+1:

**Comment by erichlof** [comment]
Oh! right - forgot to add that.  I will put something together soon.  Thanks!

**Comment by erichlof** [comment]
Hi @mrdoob ,
I added a simple example to demonstrate ShadowMesh under directional lighting and point lighting with various rotating/translating objects.
Thanks! :)
-Erich

**Comment by mrdoob** [comment]
Sweet! Many thanks! ^^

**Comment by erichlof** [comment]
@mrdoob 
Thanks for merging!  Glad to contribute.  I'm already using this feature in one of my three.js games that I just started working on!  :)

**Pull request merged**